server/http.go

```diff
@@ -264,18 +283,44 @@ func (h *httpCache) CacheHandler(w http.ResponseWriter, r *http.Request) {
 			return
 		}
 
-		rc := r.Body
+		zstdCompressed := false
+
+		// Content-Encoding must be one of "identity", "zstd" or not present.
+		ce := r.Header.Get("Content-Encoding")
+		if ce == "zstd" {
+			zstdCompressed = true
+		} else if ce != "" && ce != "identity" {
+			msg := fmt.Sprintf("Unsupported content-encoding: %q", ce)
+			http.Error(w, msg, http.StatusBadRequest)
+			h.errorLogger.Printf("PUT %s: %s", path(kind, hash), msg)
+			return
+		}
+
+		var rdr io.Reader = r.Body
 		if h.validateAC && kind == cache.AC {
 			// verify that this is a valid ActionResult
 
-			data, err := ioutil.ReadAll(rc)
+			data, err := ioutil.ReadAll(rdr)
 			if err != nil {
 				msg := "failed to read request body"
 				http.Error(w, msg, http.StatusInternalServerError)
 				h.errorLogger.Printf("PUT %s: %s", path(kind, hash), msg)
 				return
 			}
 
+			if zstdCompressed {
```

Comment:
Can kind == cache.AC be compressed?

Reply:
The REAPI proposal only covers compression using the bytestream API which is only used for CAS blobs, because it's the low-hanging fruit.

It's possible to use gRPC level compression with GetActionResult requests, in which case the data is returned in compressed form, but the data needs to be provided in uncompressed form to the gRPC bindings so I chose to keep them in uncompressed form on disk.

We could choose to support compression for AC blobs via HTTP, but I doubt there would be much to gain: they're mostly composed of high-entropy SHA256 hashes (though in string form). Others have asked before if REAPI should switch to use the binary representation of hashes instead of strings, which would be simpler than compression and also compact. I think it's fine to ignore for now though and just return uncompressed AC results.

Reply:
Yes, agree.

What I tried to ask was: Is it intentional that zstdCompressed is allowed in a code path where kind == cache.AC?

Reply:
Ah, right- this is a PUT (github's UI doesn't give much context). I think it makes sense to allow compressed PUTs, because OutputFile data can be inlined, and then it's one less special case to be documented.

Reply:
Good point! Agree.